lib/_stream_writable.js

```diff
@@ -590,6 +591,8 @@ Writable.prototype.end = function(chunk, encoding, cb) {
   // Ignore unnecessary end() calls.
   if (!state.ending)
     endWritable(this, state, cb);
+  else if (typeof cb === 'function')
```

Comment:
I think checking only on state.ending might not be enough, as the stream could be ending, but it could not have emitted ‘finish’ just yet. _final could be defined and asynchronous.

Reply:
As implemented/discussed. The second end call is always just an error and doesn’t care about ‘finish’. Should I change the semantics?

Reply:
In the docs:

> A stream method was called that cannot complete because the stream was
ended using `stream.end()`.

The problem with this sentence is that the stream might _not_ be already ended. An `end()` call was received, and the stream is in the process of _ending_, but it has not ended yet. As a result, the second callback to `end()` would be executed _before_ the first.